# Laboratorio de regresión - 3

## Significancia de factores

|                |   |
:----------------|---|
| **Nombre**      Paola Andrea Figueroa Álvarez|   |
| **Fecha**      |25/08/2025    |
| **Expediente**  751310 |   |

Descarga el archivo de publicidad y carga los datos (Advertising.csv).

In [1]:
import pandas as pd
from sklearn.linear_model import LinearRegression
import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import r2_score

In [3]:
import pandas as pd
data=pd.read_csv("Advertising.csv")
data.head()

Unnamed: 0     TV  radio  newspaper  sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9

**¿Hay alguna relación entre el presupuesto para publicidad y las ventas?**

Nuestra primera meta debe ser determinar si hay evidencia en los datos de que haya una asociación entre estas variables.

- ¿Por qué? ¿Qué resultaría si nos diéramos cuenta de la falta de relación entre el presupuesto de publicidad y las ventas?

Al analizar la relación entre el presupuesto en publicidad y las ventas podemos identificar si el gasto en marketing realmente contribuye al crecimiento de ingresos, permitiéndonos justificar si es una buena manera de distinar o no los recursos. Si no hubiera relación, significaría que aumentar el gasto en publicidad no impacta directamente en las ventas, lo que básicamente significa que una variable no depende de la otra.

**¿Qué tan fuerte es esta relación?**
Asumiendo que existe esta relación, ¿nos sirve conocer el impacto que tiene invertir en publicidad en las ventas?

La fuerza de la relación indica qué tan dependientes son las ventas del presupuesto en publicidad. En este caso tenemos una fuerte correlación, lo cual indica que si hay pequeñas variaciones en la publicidad, habrán grandes impactos en las ventas.

**¿Cuáles medios están asociados con las ventas? ¿Qué tan grande es la asociación entre un medio específico y las ventas?**

Hay 3 medios distintos en los datos. ¿Sirve invertir en los 3? ¿Conviene más invertir sólo en uno?

**¿Qué tan seguros estamos de que podríamos predecir ventas futuras?**

**¿La relación es lineal?**

**¿Hay sinergia entre estos medios?**

Puede ser que gastar \\$50,000 en publicidad y otros \\$50,000 en radio es mejor opción que gastar \\$100,000 en televisión. A esto le llamamos efecto de interacción.

Podemos usar regresión lineal para responder todas estas preguntas.

Realiza una regresión lineal:

$$ \text{ventas} \approx \beta_0 + (\beta_1)(\text{TV})$$

In [7]:
x1=data.TV.values.reshape([-1,1])
x2=data.radio.values.reshape([-1,1])
x3=data.newspaper.values.reshape([-1,1])
y=data.sales

In [9]:
# B1 de Tele
lr=LinearRegression()
lr.fit(x1,y)
lr.intercept_, lr.coef_
B0_tv= lr.intercept_
Beta1_tv=lr.coef_
B0_tv, Beta1_tv

(7.0325935491276885, array([0.04753664]))

### Verificando la precisión de nuestros coeficientes estimados

Recuerda que en el mundo real hay ruidos y errores de medición. Siempre se asume que la verdadera relación entre $X$ y $Y$ es $$Y = \beta_0 + \beta_1 X + \epsilon$$

Se asume que el término de error es independiente de $X$ (el error siempre es el mismo sin importar el valor de $X$). Este modelo describe a la *línea de regresión de la población*, que es la mejor aproximación de la verdadera relación entre $X$ y $Y$. Cuando usamos mínimos cuadrados encontramos la *línea de mínimos cuadrados*.

¿Cuál es la diferencia entre población y muestra?

La población involucra a todos los individuos o factores sobre los cuales se va a investigar, mientras que la muestra es una parte de la población, cuyo uso normalmente es para hacer inferencias respecto a que tan representativo o no es la muestra en base a la población. 

¿Cuál crees que sea la diferencia entre hacer una regresión con todos los datos de la población y una muestra de ella?

Con toda la población, los coeficientes obtenidos son los valores verdaderos, mientras que con una muestra, los coeficientes son estimaciones sujetas a un error, pues no estamos usando a todos los  individuos, por lo que los resultados podrían ser meras aproximaciones. Es por esto que al hacer inferencias se usan intervalos de confianza y pruebas de hipótesis. 

La línea de regresión de la población no se puede observar. El concepto de comparar estas líneas es una extensión natural del acercamiento estadístico estándar de usar información de una muestra para estimar características de una población grande.

Imagina que quieres encontrar la altura promedio de un mexicano $\mu$. Medir a todos y cada uno de los mexicanos en situaciones similares, con la misma regla, mismo operador, y otras incontables formas de minimizar la variación de la medida es una tarea imposible. Lo que podemos asumir es que $\hat{\mu} = \bar{y}$. La media poblacional y la media muestral son diferentes, pero la media muestral es usualmente un buen estimado.

De la misma manera, como no contamos con el 100% de la información para hacer una regresión, los coeficientes $\beta_0$ y $\beta_1$ son desconocidos. Podemos estimarlos usando mínimos cuadrados, encontrando $\hat{\beta_0}$ y $\hat{\beta_1}$. Puede que las muestras que tengamos en ese momento estén un poco por encima de la media, pero otras muestras en otro momento puede que estén debajo de la media. En general, esperamos que el promedio de las aproximaciones $\hat{\mu}$ aproxime a $\mu$.

Esto lleva a la pregunta: ¿qué tan cercanos son nuestros coeficientes estimados a los verdaderos coeficientes? Utilizamos el concepto de error estándar para evaluar esto.

$$ \text{Var}(\hat{\mu})=\text{SE}(\hat{\mu})^2 = \frac{\sigma^2}{n} $$

Donde $\sigma$ es la desviación estándar de cada una de las observaciones $y_i$ de $Y$. El error estándar nos dice la cantidad promedio que el estimado difiere del valor verdadero. Podemos ver en la fórmula que entre más observaciones tengamos el error se hace más pequeño. Las fórmulas para errores estándar de $\hat{\beta_0}$ y $\hat{\beta_1}$ son:

$$ \text{SE}(\hat{\beta_0})^2 = \sigma^2 [\frac{1}{n} + \frac{\bar{x}^2}{\sum_{i=1}^n (x_i - \bar{x})^2}]$$

$$ \text{SE}(\hat{\beta_1})^2 = \frac{\sigma^2}{\sum_{i=1}^n (x_i - \bar{x})^2}$$

$$ \sigma^2 = \text{Var}(\epsilon) = \text{RSE}^2 = \frac{\text{RSS}}{n-p}$$

Para que estas fórmulas sean validas asumimos que los errores $\epsilon_i$ tienen varianza común $\sigma^2$ y que no están correlacionados.

Calcula los errores estándar de los coeficientes

In [12]:
n=len(y)
p=2
x1_mean = np.mean(x1)
y_mean = np.mean(y)
RSS=sum((lr.predict(x1)-y)**2)
RSS

2102.5305831313512

In [14]:
sigma2= RSS/(n-p)
sigma2

10.618841328946218

In [16]:
SE_B11= sigma2/sum((x1-x1_mean)**2)
SE_B1=np.sqrt(SE_B11)
SE_B1

array([0.00269061])

In [18]:
SE_B0=sigma2*(1/n + ((x1_mean**2)/sum((x1-x1_mean)**2)))
SE_B0=np.sqrt(SE_B0)
SE_B0

array([0.45784294])

Estos errores se pueden usar para calcular intervalos de confianza. Un intervalo de confianza del $95\%$ se define como un rango de valores en el cuál se encuentra el desconocido valor verdadero con un $95\%$ de probabilidad.

Otra forma de verlo es que si tomamos muestras repetidas y construimos un intervalo de confianza para cada una, el $95\%$ de los intervalos creados van a contener el valor verdadero. Para la regresión el intervalo de confianza del $95\%$ toma la forma:

$$ \hat{\beta_j} \pm 2\text{SE}(\hat{\beta_j})$$

Calcula los intervalos de confianza para los coeficientes estimados:

In [20]:
IntervaloB0_up= B0_tv+2*SE_B0
IntervaloB0_low= B0_tv-2*SE_B0
IntervaloB0_up, IntervaloB0_low

(array([7.94827943]), array([6.11690767]))

In [26]:
IntervaloB1_up= Beta1_tv+2*SE_B1
IntervaloB1_low= Beta1_tv-2*SE_B1
IntervaloB1_up, IntervaloB1_low

(array([0.05291785]), array([0.04215543]))

Los errores estándar también se usan para realizar pruebas de hipótesis. La prueba de hipótesis más común es probar la hipótesis nula de:

$$ H_0: \text{No hay relación entre } X \text{ y } Y \ \ \ \ (\beta_1=0)$$

contra la hipótesis alternativa:
$$ H_0: \text{Hay alguna relación entre } X \text{ y } Y \ \ \ (\beta_1 \neq 0)$$

Explica con tus palabras el significado de la hipótesis nula y la hipótesis alternativa.

La hipótesis nula es una afirmación que se toma como verdadera hasta que estadísticamente se pruebe lo contrario. La hipótesis alternativa es lo contrario de la nula: representa lo que queremos demostrar, ya sea que hay una relación, diferencia o efecto. 

Para probar la hipótesis nula debemos determinar si nuestro estimado $\hat{\beta_1}$ de $\beta_1$ está lo suficientemente alejado de cero para que podamos decir con confianza que este valor no es cero. 

¿Qué tan lejos? Depende de qué tanta confianza tengamos en el estimado encontrado. Si nuestro error estándar es pequeño y nuestro estimado está alejado de cero podríamos decir que hay muy poca probabilidad de que el valor verdadero sea 0. En cambio, si nuestro error estándar es grande y nuestro estimado está muy cerca de cero, entonces podrías ser que el valor verdadero sea cero y que no haya relación entre las variables.

Se calcula un *estadístico t* dado por
$$ t = \frac{\hat{\beta_j} - \mu}{\text{SE}(\hat{\beta_j})} $$

donde $\mu$ es el valor contra el que queremos probar.

Calcula el estadístico t para tus coeficientes estimados, usando como referencia la prueba de hipótesis.

#### Hipótesis nula y alternativa 
En este caso miu es 0 porque la hipotesis es que sea igual o diferente de 0
#### Ho= B0= 0
#### Hi= B0 ≠ 0

In [28]:
t0= B0_tv/SE_B0
t0

array([15.36027517])

In [30]:
t1= Beta1_tv/SE_B1
t1

array([17.6676256])

La distribución t tiene forma de campana y se parece bastante a la distribución normal cuando $n > 30$. Ya sólo es cuestión de calcular la probabilidad de observar cualquier número tal que su valor absoluto sea igual o mayor que el valor absoluto del estadístico t calculado. En otras palabras:
$$ P(|x| \geq |t|) $$

A esta probabilidad la llamamos *p-value*. Un *p-value* pequeño indica que es poco probable que exista por puro azar una relación significativa entre predictor y respuesta, en caso de que no haya una asociación real entre predictor y respuesta. En otras palabras, el *p-value* te dice la probabilidad de que parezca que hay relación cuando no la hay.

Si el *p-value* es pequeño, inferimos que sí hay una asociación entre el predictor y la respuesta, y **rechazamos la hipótesis nula**.
  

¿Qué tan pequeño? Depende de la aplicación. Un valor muy común es del $5\%$.

Utiliza el siguiente código para calcular el *p-value* para tus coeficientes

`from scipy import stats`

`p_bj = 2*(1 - stats.t.cdf(np.abs(t_bj), n-p))`

In [32]:
from scipy import stats

In [34]:
p_b1 = 2*(1 - stats.t.cdf(np.abs(t1), df=n-2))
p_b1

array([0.])

In [36]:
p_b0 = 2*(1 - stats.t.cdf(np.abs(t0), df=n-2))
p_b0

array([0.])

¿Se rechaza la hipótesis nula? ¿Qué significa?

Para el caso de las televisiones, podemos decir que el coeficiente de radio es positivo y estadísticamente significativo, con un intervalo de confianza que no incluye al cero, un estadístico t elevado y un p-value menor a 0.05. Esto nos dice que sí se rechaza la hipótesis nula, pues la que calculamos es muy acercada a la real. En este caso podemos decir que hay suficiente evidencia estadística para decir que el coeficiente es distinto de cero y, por tanto, la variable sí influye en las ventas.

Realiza otras dos regresiones. Ya tienes hecha la regresión de ventas dado el gasto en publicidad de TV. Realiza la regresión para gastos en radio y gastos en periódico. Organiza las respuestas para que debajo de esta celda se tenga:
- Título de regresión
- Coeficientes estimados
- Errores estándar de los coeficientes
- Intervalos de confianza
- Estadísticos t
- p-values
- Observaciones

## Regresión de ventas de radio

### Coeficientes

In [38]:
lr2=LinearRegression()
lr2.fit(x2,y)
lr2.intercept_, lr2.coef_
B0_radio= lr2.intercept_ 
Beta1_radio=lr2.coef_
B0_radio, Beta1_radio

(9.311638095158285, array([0.20249578]))

### Errores estándar de los coeficientes

In [40]:
n=len(y)
p=2
x2_mean = np.mean(x2)
y_mean = np.mean(y)
RSS2=sum((lr2.predict(x2)-y)**2)
RSS2

3618.4795490250876

In [42]:
sigma2_2= RSS2/(n-p)
sigma2_2

18.275149237500443

In [44]:
SE_B12= sigma2_2/sum((x2-x2_mean)**2)
SE_B1_2=np.sqrt(SE_B12)
SE_B1_2

array([0.02041131])

In [46]:
SE_B0_2=sigma2_2*(1/n + ((x2_mean**2)/sum((x2-x2_mean)**2)))
SE_B0_2=np.sqrt(SE_B0_2)
SE_B0_2

array([0.5629005])

### Intervalos de confianza

In [48]:
IntervaloB0_up_radio= B0_radio+2*SE_B0_2
IntervaloB0_low_radio= B0_radio-2*SE_B0_2
IntervaloB1_up_radio= Beta1_radio+2*SE_B1_2
IntervaloB1_low_radio= Beta1_radio-2*SE_B1_2
IntervaloB0_up_radio, IntervaloB0_low_radio,IntervaloB1_up_radio,IntervaloB1_low_radio

(array([10.43743909]),
 array([8.1858371]),
 array([0.2433184]),
 array([0.16167317]))

### Estadísticos t

In [50]:
t0_radio= B0_radio/SE_B0_2
t0_radio
t1_radio= Beta1_radio/SE_B1_2
t1_radio
t0_radio, t1_radio

(array([16.54224531]), array([9.92076547]))

### p_value

In [52]:
p_b1_radio = 2*(1 - stats.t.cdf(np.abs(t1_radio), df=n-2))
p_b1_radio
p_b0_radio = 2*(1 - stats.t.cdf(np.abs(t0_radio), df=n-2))
p_b0_radio
p_b1_radio, p_b0_radio

(array([0.]), array([0.]))

### Observaciones

El coeficiente de radio es positivo y estadísticamente significativo, con un intervalo de confianza que no incluye al cero, un estadístico t elevado y un p-value menor a 0.05. Estos resultados permiten rechazar la hipótesis nula y afirmar que la inversión en radio tiene un efecto real y relevante en el incremento de las ventas.

## Regresión de ventas de periódico

### Coeficientes

In [54]:
lr3=LinearRegression()
lr3.fit(x3,y)
lr3.intercept_, lr3.coef_
B0_paper= lr3.intercept_ 
Beta1_paper= lr3.coef_
B0_paper, Beta1_paper

(12.35140706927816, array([0.0546931]))

### Errores estándar de los coeficientes

In [56]:
n=len(y)
p=2
x3_mean = np.mean(x3)
y_mean = np.mean(y)
RSS3= np.sum((lr3.predict(x3)-y)**2)
RSS3

5134.804544111939

In [58]:
sigma2_3= RSS3/(n-p)
sigma2_3

25.933356283393632

In [60]:
SE_B3= sigma2_3/sum((x3-x3_mean)**2)
SE_B3=np.sqrt(SE_B3)
SE_B3

array([0.01657572])

In [62]:
SE_B0_3=sigma2_3*(1/n + ((x3_mean**2)/sum((x3-x3_mean)**2)))
SE_B0_3=np.sqrt(SE_B0_3)
SE_B0_3

array([0.62142019])

### Intervalos de confianza

In [64]:
IntervaloB0_up_paper= B0_paper+2*SE_B0_3
IntervaloB0_low_paper= B0_paper-2*SE_B0_3
IntervaloB1_up_paper= Beta1_paper+2*SE_B3
IntervaloB1_low_paper= Beta1_paper-2*SE_B3
IntervaloB0_up_paper, IntervaloB0_low_paper,IntervaloB1_up_paper,IntervaloB1_low_paper

(array([13.59424744]),
 array([11.10856669]),
 array([0.08784454]),
 array([0.02154165]))

### Estadísticos t

In [66]:
t0_paper= B0_paper/SE_B0_3
t0_paper
t1_paper= Beta1_paper/SE_B3
t1_paper
t0_paper, t1_paper

(array([19.87609562]), array([3.29959074]))

### p_value

In [68]:
p_b1_paper = 2*(1 - stats.t.cdf(np.abs(t1_paper), df=n-2))
p_b1_paper
p_b0_paper = 2*(1 - stats.t.cdf(np.abs(t0_paper), df=n-2))
p_b0_paper
p_b1_paper, p_b0_paper

(array([0.0011482]), array([0.]))

### Observaciones

En el caso de periódico, el coeficiente estimado es positivo y su intervalo de confianza incluye al cero, acompañado de un estadístico t pequeño y un p-value menor a 0.05. A diferencia de los otros casos, el p-value de B1 dio diferente, lo que nos dice que sí importa lo que hagamos con el periódico, indicando que sí se rechaza la hipótesis nula porque se afirma que sí es diferente de 0 aunque no es tan seguro o significativo como la radio o tv.

## Regresión lineal múltiple

En lugar de hacer una regresión para cada factor independiente, quizás se puede extender el modelo para que tenga varios factores dentro:

$$ Y = \beta_0 + \beta_1 X_1 + \beta_2 X_2 + ... + \beta_p X_p + \epsilon $$

Para nuestro ejemplo de publicidad:

$$ \text{sales} = \beta_0 + \beta_1 (\text{TV}) + \beta_2 (\text{radio}) + \beta_3 (\text{newspaper}) + \epsilon $$

Utiliza la librería `statsmodels` para realizar la regresión. Por defecto la librería statsmodels no toma en cuenta el intercepto ($\beta_0$), por lo que se tendrá que agregar una columna de unos de tamaño *n* a la matriz X.

`import statsmodels.api as sm`

`ols = sm.OLS(Y, X)`

`results = ols.fit()`

`results.summary()`

In [71]:
import statsmodels.api as sm

In [73]:
n=len(y)
unos=np.ones([n,1])
X=np.hstack([unos, x1,x2,x3])
# X es la concatenación de todos los factores
ols= sm.OLS(y,X)
results= ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Sat, 30 Aug 2025   Prob (F-statistic):           1.58e-96
Time:                        12:22:09   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.000       2.324       3.554
x1             0.0458      0.001     32.809      0.000       0.043       0.049
x2             0.1885      0.009     21.893      0.000       0.172       0.206
x3            -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

¿Qué diferencias puedes observar entre los *p-values* de una regresión múltiple y los encontrados en las regresiones simples? ¿Por qué crees que existen estas diferencias?

Los resultados de los p-values son iguales para B0, tele y radio, sin embargo, para los periódicos arroja un p-value alto, lo que significa que para este caso, la hipótesis nula no se rechaza, por lo que hay suficiente evidencia estadística para afirmar que no hay una relación con esta variable y, por lo tanto, es irrelevante; su información ya está en alguna de las otras variables, es decir la combinación entre tv y radio ya te da lo del periódico.
La correlación no es igual a causa, lo que significa que en una regresión simple las variables, al analizarlas por sí solas, pueden mostrar asociación con otras variables. Sin embargo, al incluirlas junto con televisión y radio en una regresión múltiple, parte de su efecto se “absorbe” por estas variables más influyentes, revelando que su aporte real es mínimo o nulo.

### Regresión sin periódico

In [75]:
n=len(y)
unos=np.ones([n,1])
x_nueva= x1*x2.reshape([-1,1])
X=np.hstack([unos, x1,x2,x_nueva])
# X es la concatenación de todos los factores
ols= sm.OLS(y,X)
results= ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.968
Model:                            OLS   Adj. R-squared:                  0.967
Method:                 Least Squares   F-statistic:                     1963.
Date:                Sat, 30 Aug 2025   Prob (F-statistic):          6.68e-146
Time:                        12:22:39   Log-Likelihood:                -270.14
No. Observations:                 200   AIC:                             548.3
Df Residuals:                     196   BIC:                             561.5
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          6.7502      0.248     27.233      0.000       6.261       7.239
x1             0.0191      0.002     12.699      0.000       0.016       0.022
x2             0.0289      0.009      3.241      0.001       0.011       0.046
x3             0.0011   5.24e-05     20.727      0.000       0.001       0.001
==============================================================================
Omnibus:                      128.132   Durbin-Watson:                   2.224
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1183.719
Skew:                          -2.323   Prob(JB):                    9.09e-258
Kurtosis:                      13.975   Cond. No.                     1.80e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.8e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

ventas= 6.75 + (0.019 + 0.0011 radio) TV + 0.028 radio

Entre mas dinero tenga en la radio, más efectivo es la televisión o viceversa

### Modelo choncho

In [114]:
n=len(y)
unos=np.ones([n,1])
x1x2= x1*x2.reshape([-1,1])
x1x3= x1*x3.reshape([-1,1])
x2x3= x2*x3.reshape([-1,1])
X=np.hstack([unos, x1,x2,x3,x1x2,x1x3,x2x3])
# X es la concatenación de todos los factores
ols= sm.OLS(y,X)
results= ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.969
Model:                            OLS   Adj. R-squared:                  0.968
Method:                 Least Squares   F-statistic:                     993.3
Date:                Thu, 28 Aug 2025   Prob (F-statistic):          3.68e-142
Time:                        17:43:24   Log-Likelihood:                -267.49
No. Observations:                 200   AIC:                             549.0
Df Residuals:                     193   BIC:                             572.1
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          6.4602      0.318     20.342      0.000       5.834       7.087
x1             0.0203      0.002     12.633      0.000       0.017       0.024
x2             0.0229      0.011      2.009      0.046       0.000       0.045
x3             0.0170      0.010      1.691      0.092      -0.003       0.037
x4             0.0011   5.72e-05     19.930      0.000       0.001       0.001
x5         -7.971e-05   3.58e-05     -2.227      0.027      -0.000   -9.12e-06
x6            -0.0001      0.000     -0.464      0.643      -0.001       0.000
==============================================================================
Omnibus:                      111.385   Durbin-Watson:                   2.222
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              767.483
Skew:                          -2.046   Prob(JB):                    2.20e-167
Kurtosis:                      11.681   Cond. No.                     3.80e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.8e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

## Referencia

James, G., Witten, D., Hastie, T., Tibshirani, R.,, Taylor, J. (2023). An Introduction to Statistical Learning with Applications in Python. Cham: Springer. ISBN: 978-3-031-38746-3

## Addendum

Para calcular los *p-values* de los parámetros sin `statsmodels`:
1. Calcular RSS
2. Calcular RSE
3. `var_beta = np.linalg.inv(X.T @ X) * rse**2` (X con columna de intercepto)
4. `std_beta = np.sqrt(var_beta.diagonal())` El orden de los valores corresponde al orden de los factores en las columnas de la matriz $X$.
5. Calcular *estadístico t*
6. Calcular *p-value*